jsonschema/exceptions.py

```diff
@@ -5,13 +5,14 @@
 from pprint import pformat
 from textwrap import dedent, indent
 import itertools
+import typing
 
 import attr
 
 from jsonschema import _utils
 
-WEAK_MATCHES = frozenset(["anyOf", "oneOf"])
-STRONG_MATCHES = frozenset()
+WEAK_MATCHES: typing.FrozenSet[str] = frozenset(["anyOf", "oneOf"])
```

Comment:
(Ignorance question, I'll have to read myself a bit more before merging) but is there a more generic type here than this? I'd like to not say this is a frozenset, in the future it may (likely will) become a `pyrsistent.pset`. So just something like `collections.abc.Set` or whatever the typing equivalent is would be a bit preferable.

Reply:
I think the best we can do is `typing.Collection[str]`, which is the type for `collections.abc.Collection`.

`typing.Set` doesn't work, as it is the type for `builtins.set`, not `collections.abc.Set`.
Most of `collections.abc` is mirrored in `typing`, and even `collections.abc.MutableSet` maps to `typing.MutableSet`. I guess `Set` was skipped because of the likelihood of confusion with the builtin set type.

In py3.9+, all of the `collections.abc` types are usable as generics, so we'd be able to write
```python
WEAK_MATCHES: collections.abc.Set[str] = ...
```
and I would expect it to work.

The big alternative, which I could apply here and elsewhere, is to use `from __future__ import annotations`, now that py3.6 has been dropped. The future import enables several features and behaviors, but the one that is relevant here is that it allows the use of builtin types in annotations. So we can do
```python
from __future__ import annotations

WEAK_MATCHES: frozenset[str] = ...
```

If you prefer the results of the future import, I can apply it to all of the places where I had to add annotations.

Reply:
Yeah let's go with the future import! After that I think I'll merge. Thanks again for the help here!

Reply:
I've just gotten around to doing this. Wherever it simplified things, I added a `from __future__ import annotations`.

Let me know if there are any issues with it, but I think we're good to go on this! 😄